src/mscorlib/src/System/Collections/Generic/PooledIEnumerator.cs

```diff
@@ -0,0 +1,34 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace System.Collections.Generic
+{
+    internal class PooledIEnumerator<TValue, TEnumerator, TArg> : IEnumerator<TValue> where TEnumerator : IEnumerator<TValue>
+    {
+        private TEnumerator _enumerator;
+
+        internal PooledIEnumerator<TValue, TEnumerator, TArg> Initalize(Func<TArg, TEnumerator> ctor, TArg state)
+        {
+            _enumerator = ctor(state);
+            return this;
+        }
+        public bool MoveNext() => _enumerator.MoveNext();
+        public TValue Current => _enumerator.Current;
+        object IEnumerator.Current => Current;
+
+        public void Dispose()
+        {
+            _enumerator = default(TEnumerator);
+            ObjectPool<PooledIEnumerator<TValue, TEnumerator, TArg>>.Shared.Return(this);
+        }
+        public void Reset()
+        {
+            throw new NotSupportedException();
```

Comment:
Should this not call into `_enumerator.Reset()`? Right now it looks breaking (on a bigger level than just the return type change) since the list and dictionary enumerators support Reset.